src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -987,8 +987,12 @@ HeapWord* ShenandoahHeap::allocate_memory(ShenandoahAllocRequest& req) {
     // is testing that the GC overhead limit has not been exceeded.
     // This will notify the collector to start a cycle, but will raise
     // an OOME to the mutator if the last Full GCs have not made progress.
-    if (result == nullptr && !req.is_lab_alloc() && get_gc_no_progress_count() > ShenandoahNoProgressThreshold) {
+    // gc_no_progress_count is incremented following each degen or full GC that fails to achieve is_good_progress().
+    // Note that Generational Shenandoah may increment no_progress_count faster than traditional Shenandoah because young
+    // GCs, which may degenerate, typically occur more frequently than single-generation Global GCs.
```

Comment:
Thanks for asking these questions and digging into the details.  I see in your other comment that it is the intention that we only increment gc_no_progress_count for consecutive full GCs.  However, our implementation seems to have drifted from that intention.  We also increment following a degen that does not make progress.  And then we increment again in the Full GC after we upgrade to full.  Maybe there are better ways to fix this.  I'm investigating.

Reply:
I'm of the "opinion" that consecutive Full GC failures means no intervening successful concurrent or degenerated collections.

Reply:
I'm proposing two changes:
1. Do not increment gc_no_progress in failed degens
2. Reset gc_no_progress to zero after successful concurrent gc

I will remove this comment, and I will test this change.  I believe this may allow me to backoff the extreme testing overrides of ShenandoahNoProgressThreshold.

Reply:
If we don't reset gc_no_progress to zero after successful concurrent gc, it seems the use of ShenandoahNoProgressThreshold might be kind of "arbitrary".  For example, TestThreadFailure repeatedly drives the heap to exhaustion, and then it cleans up and does it again.  Each time it reaches exhaustion, we escalate to FullGC and throw OOM, which causes the mutator to release its memory.  Having released its memory, we'll run multiple successful concurrent GCs until we once again have exhausted memory, at which time we'll once again escalate to FullGC, which will be unproductive and we'll thrown OOM.

The sixth time though this loop (the sixth consecutive Full GC with many successful intervening concurrent GCs), we'll see that _gc_no_progress_count exceeds ShenandoahNoProgressThreshold and will throw OOM without even trying to do a Full GC.  I think this is not what we want.